### PR TITLE
add schema-based trnasforms

### LINKED ISSUE #22501: [Tour of Beam] [Task] Learning content for "Schema-based Transforms" module
### What needs to happen?
Create learning content and challenge for Tour of Beam::Introduction module consisting of following unit:

- CoGroup
- Join
- Filter
- Select
- New motivating example/challenge

### Issue Priority
Priority: 3

### PR DESCRIPTION
**Pull request by sirenbyte**
fixes #22501
**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by sirenbyte** [comment]
R: @olehborysevych

**Review by alxp1982 — CHANGES_REQUESTED** [review]

File: learning/tour-of-beam/learning-content/java/schema-based-transforms/coder/description.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+# Coder
+
+A Coder<T> defines how to encode and decode values of type T into byte streams.
```
Comment (alxp1982) [inline learning/tour-of-beam/learning-content/java/schema-based-transforms/coder/description.md:17]: When Beam runners execute your pipeline, they often need to convert an object to and from a binary format. The Beam SDKs use objects called Coders to perform coding and decoding. 

In Beam SDK for Java, the base type `Coder` contains conversion methods. SDK also contains Coder subclasses that work with various standard Java types, such as Integer, Long, Double, StringUtf8, and more. You can find all the available Coder subclasses in the [Coder package](https://github.com/apache/beam/tree/master/sdks/java/core/src/main/java/org/apache/beam/sdk/coders).

File: learning/tour-of-beam/learning-content/java/schema-based-transforms/coder/description.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+Coder instances are serialized during job creation and deserialized before use. This will generally be performed by serializing the object via Java Serialization.
+
+Coder classes for compound types are often composed of coder classes for types contains therein. The composition of Coder instances into a coder for the compound class is the subject of the Coder Provider type, which enables automatic generic composition of Coder classes within the CoderRegistry. See Coder Provider and CoderRegistry for more information about how coders are inferred.
+
+When you write your Custom code, you inherit from Coder and you have to implement its methods.
```
Comment (alxp1982) [inline learning/tour-of-beam/learning-content/java/schema-based-transforms/coder/description.md:23]: When you create custom objects and schemas, you need to create a subclass of `Coder` for your object and implement the following methods:

File: learning/tour-of-beam/learning-content/java/schema-based-transforms/coder/description.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+* `encode` - converting objects to bytes
+* `decode` - converting bytes to objects
+* `getCoderArguments` - If it is a `Coder` for a parameterized type, returns a list of `Coders` used for each of the parameters, in the same order in which they appear in the type signature of the parameterized type.
+* `verifyDeterministic` - throw the `Coder.NonDeterministicException`, if the encoding is not deterministic.
+
+Entity `VendorToPassengerDTO` to work with pipeline:
```
Comment (alxp1982) [inline learning/tour-of-beam/learning-content/java/schema-based-transforms/coder/description.md:29]: For example, consider the following schema type:

File: learning/tour-of-beam/learning-content/java/schema-based-transforms/coder/description.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+    @Override
+    public void verifyDeterministic() {
+    }
+}
+```
+
+### Playground exercise
+
+You can find the complete code of this example in the playground window you can run and experiment with.
```
Comment (alxp1982) [inline learning/tour-of-beam/learning-content/java/schema-based-transforms/coder/description.md:116]: Excercise description and a small challenge to invite user to experiment.

File: learning/tour-of-beam/learning-content/java/schema-based-transforms/coder/description.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+One of the differences you will notice is that it also contains the part to output `PCollection` elements to the console.
```
Comment (alxp1982) [inline learning/tour-of-beam/learning-content/java/schema-based-transforms/coder/description.md:118]: Remove

File: learning/tour-of-beam/learning-content/java/schema-based-transforms/coder/description.md [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+Do you also notice in what order elements of PCollection appear in the console? Why is that? You can also run the example several times to see if the output stays the same or changes.
```
Comment (alxp1982) [inline learning/tour-of-beam/learning-content/java/schema-based-transforms/coder/description.md:120]: Remove

**Review by alxp1982 — CHANGES_REQUESTED** [review]

File: learning/tour-of-beam/learning-content/java/schema-based-transforms/filter/example/Task.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            this.cost = cost;
+            this.transactionDuration = transactionDuration;
+        }
+    }
+
+    public static void main(String[] args) {
+        PipelineOptions options = PipelineOptionsFactory.fromArgs(args).create();
+        Pipeline pipeline = Pipeline.create(options);
+
+        UserPurchase user1 = new UserPurchase(1L, "America", 123, 22);
```
Comment (alxp1982) [inline learning/tour-of-beam/learning-content/java/schema-based-transforms/filter/example/Task.java:70]: UserPurchase user1 = new UserPurchase(1L, "USA", 123, 22);

File: learning/tour-of-beam/learning-content/java/schema-based-transforms/filter/description.md [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+```
+PCollection<UserAccount> users = readUsers();
+users.apply(Filter
+    .whereFieldNames(Lists.newArrayList("spendOnBooks", "spendOnMovies"),
+        row -> return row.getDouble("spendOnBooks") + row.getDouble("spendOnMovies") > 100.00));
+```
+
+### Playground exercise
+
+You can find the complete code of this example in the playground window you can run and experiment with.
```
Comment (alxp1982) [inline learning/tour-of-beam/learning-content/java/schema-based-transforms/filter/description.md:64]: In the playground window, you can find an example of `Filter` usage to find all user purchase transactions that took more than 50 seconds to complete. You can run and experiment with it. Try to filter input by several fields to find, for example, all transactions over $100 that happened in France.

**Review by alxp1982 — CHANGES_REQUESTED** [review]

File: learning/tour-of-beam/learning-content/java/schema-based-transforms/group/description.md [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+              Field.of("transactionDurations", FieldType.array(FieldType.INT64)));
+```
+
+The result will be a new row schema containing the fields `total_cost`, ``top_purchases``, and `transactionDurations`, containing the sum of all purchases costs (for that user and country), the top ten purchases, and a histogram of transaction durations. The schema will also contain a key field, which will be a row containing userId and country.
+
+Note that usually the field type can be automatically inferred from the `Combine.CombineFn` passed in. However sometimes it cannot be inferred, due to Java type erasure, in which case a `Schema.Field` object containing the field type must be passed in. This is currently the case for `ApproximateQuantilesCombineFn` in the above example.
+
+### Playground exercise
+
+You can find the complete code of this example in the playground window you can run and experiment with.
```
Comment (alxp1982) [inline learning/tour-of-beam/learning-content/java/schema-based-transforms/group/description.md:57]: In the playground window, you can find an example you can run and experiment with.  This example illustrates `Group` transform usage to group purchases by the user together with aggregation to calculate the total value of purchases made. 

Can you modify it to get total value of all purchases made by user in different countries?

File: learning/tour-of-beam/learning-content/java/schema-based-transforms/group/example/Task.java [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+            this.cost = cost;
+            this.transactionDuration = transactionDuration;
+        }
+    }
+
+    public static void main(String[] args) {
+        PipelineOptions options = PipelineOptionsFactory.fromArgs(args).create();
+        Pipeline pipeline = Pipeline.create(options);
+
+        UserPurchase user1 = new UserPurchase(1L, "America", 123, 22);
```
Comment (alxp1982) [inline learning/tour-of-beam/learning-content/java/schema-based-transforms/group/example/Task.java:71]: UserPurchase user1 = new UserPurchase(1L, "USA", 123, 22);

File: learning/tour-of-beam/learning-content/java/schema-based-transforms/group/example/Task.java [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        UserPurchase user2 = new UserPurchase(1L, "Brazilian", 645, 86);
```
Comment (alxp1982) [inline learning/tour-of-beam/learning-content/java/schema-based-transforms/group/example/Task.java:72]: UserPurchase user2 = new UserPurchase(1L, "Brazil", 645, 86);

**Review by alxp1982 — CHANGES_REQUESTED** [review]

File: learning/tour-of-beam/learning-content/java/schema-based-transforms/coder/description.md [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+
+    @Override
+    public void verifyDeterministic() {
+    }
+}
+```
+
+### Playground exercise
+
+In the playground window you can find examples of using `Coder`. By running this example, you will see user statistics in certain games. Can you change its structure so that an **additional field** appears if the point is greater than **10** to supply `true`.
```
Comment (alxp1982) [inline learning/tour-of-beam/learning-content/java/schema-based-transforms/coder/description.md:126]: Could you pleas re-phrase the user challenge? I'm not sure what we ask user to do

**Pull request closed without merging**